Add to_dataframe() to Record.

Ivorforce opened the pull request:
Convenient for a quick start with records.

tompollard reviewed (commented):
Looks good to me, but I'll leave @bemoody and/or @cx1111 to comment too.

tompollard commented:
> It doesn't seem very efficient to store a regularly sampled signal as a dataframe. What's the use-case for this?

Maybe not efficient, but lots of people are familiar with Pandas DataFrames which is good for accessibility. This small chunk of code allows people to quickly play around with the data in a familiar structure. e.g. perhaps not a great example of a use case, but I could do something like:

View the data:

```python
import wfdb
import matplotlib.pyplot as plt
import seaborn as sns

record = wfdb.rdrecord('sample-data/a103l')
df = record.to_dataframe()
df.head()

                              II         V     PLETH
0 days 00:00:00        -0.023596  0.867586  0.482203
0 days 00:00:00.004000 -0.036981  0.982985  0.544374
0 days 00:00:00.008000 -0.062923  0.859791  0.478212
0 days 00:00:00.012000 -0.092452  0.788783  0.442857
0 days 00:00:00.016000 -0.094522  0.851996  0.474302
```

Plot the distribution of measurements:

```python
sns.boxplot(data=df)
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/822601/171472428-b475d859-c5fd-4cf2-931b-5ffd0b2f15bd.png)

cx1111 commented:
Ok, makes sense. @Ivorforce can you please add a unit test and an example in the demo notebook?

Ivorforce commented on wfdb/io/record.py:
```diff
@@ -851,6 +851,35 @@ def _arrange_fields(self, channels, sampfrom, smooth_frames):
+        A dataframe, with sig_name in the columns. The index is a DatetimeIndex
+        if both base_date and base_time were set, otherwise a TimedeltaIndex.
+        """
+        if self.base_datetime is not None:
```
In this case it probably doesn't matter, but I generally find `is not None` to be less ambiguous and error-prone.

Ivorforce commented on wfdb/io/record.py:
```diff
@@ -851,6 +851,35 @@ def _arrange_fields(self, channels, sampfrom, smooth_frames):
+            )
+
+        return pd.DataFrame(
+            data=self.p_signal,
```
Perhaps. I added a variable check to support all 4 combinations of data. I'm not sure if it makes sense though so I'll let @cx1111 decide whether we want to keep it. The real 'workable data' is p_signal (or maybe e_p_signal) anyway.

Ivorforce commented:
Yes, sorry for not providing motivation - @tompollard got it exactly right. I opened this PR for accessibility to new people and existing libraries.

Editing an ipython notebook is a bit of a struggle if you don't want to change metadata, but I added a unittest and a demo like asked. I also added support for digital and expanded signals. Let me know if it's ready like this.

bemoody commented:
This looks nice!  I don't really use Pandas so I don't have much to add here.  One question that comes to mind: is there a way for the dataframe to indicate the physical units of each channel/column?

Ivorforce commented:
No support for units in dataframes. See this issue: https://github.com/pandas-dev/pandas/issues/10349

Ivorforce commented:
I switched to `pd.Timedelta()` because it has a higher precision than `datetime.timedelta`, resulting in a lower probability of precision loss for the index.

tompollard commented:
@Ivorforce it looks like one of the tests is failing. Please could you take a look when you have time?

```
_________________________ TestRecord.test_to_dataframe _________________________
[18](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:19)
[19](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:20)
self = <tests.test_record.TestRecord testMethod=test_to_dataframe>
[20](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:21)
[21](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:22)
    def test_to_dataframe(self):
[22](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:23)
        record = wfdb.rdrecord("sample-data/test01_00s")
[23](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:24)
        df = record.to_dataframe()
[24](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:25)
[25](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:26)
        self.assertEqual(record.sig_name, list(df.columns))
[26](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:27)
        self.assertEqual(len(df), record.sig_len)
[27](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:28)
>       self.assertEqual(df.index[0], pd.Timedelta())
[28](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:29)
[29](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:30)
tests/test_record.py:538: 
[30](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:31)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[31](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:32)
[32](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:33)
>   ???
[33](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:34)
E   ValueError: cannot construct a Timedelta without a value/unit or descriptive keywords (days,seconds....)
[34](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:35)
[35](https://github.com/MIT-LCP/wfdb-python/runs/6904057193?check_suite_focus=true#step:5:36)
pandas/_libs/tslibs/timedeltas.pyx:986: ValueError
```

Ivorforce commented:
Whoops, looks like I missed this. Tests should pass now.

cx1111 commented:
Ty!
